README


This archive contains the distribution Basket-Calc,
version 0.4:

  Basket/Cart calculation library with support for currency conversion, discounts and tax

This software is Copyright (c) 2013 by Tobias Kirschstein.

This is free software, licensed under:

  The (three-clause) BSD License




Changes
==================================================
Changes from 2012-12-09 00:00:00 +0000 to present.
==================================================

----------------------------------------
version 0.4 at 2013-12-09 20:54:18 +0000
----------------------------------------

  Change: 17f93e9eff3aa4a72016bb9a7a6ef7741494560b
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-12-10 09:53:56 +0000

    perl 5.18 changes, only print in debug 

----------------------------------------
version 0.3 at 2013-05-08 03:15:20 +0000
----------------------------------------

  Change: 1599b6a27432db44c80bd0e3fcf3ac0ce579a2fe
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-08 15:15:20 +0000

    fixed rounding for some perl versions 

  Change: 271a5607f477dd07f5fe5b96020530798318d3b2
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-03 04:18:54 +0000

    create GPG signed tags 

  Change: b09d06e745494b6992f1c8a353b565cbdd20c7e1
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-03 04:14:15 +0000

    v0.2 

----------------------------------------
version 0.2 at 2013-05-02 16:13:46 +0000
----------------------------------------

  Change: fcf819d649ee9ca5c1240474301751e209a140f3
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-03 04:13:46 +0000

    got versioning working with Perl::Critic 

  Change: 8e929b1a41b80af8d3ca52c7682ca81cf5e34b43
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-03 03:33:28 +0000

    removed VERSION stufff, it's auto generated and inserted 

  Change: 5fdb15bf4978b5383900c3764c4ad1164263ffe4
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-03 02:52:21 +0000

    fixed pod syntax 

  Change: 79d25b2f3f777c1282bf306b504930dbb9ba0b07
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-03 02:47:08 +0000

    fixed even more pod stuff for dzil 

  Change: cef673646d90af4a3c2cfab131bf09d9bc58acb9
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-03 02:15:58 +0000

    fixed versioning 

  Change: aaf2729274f712d02dc5014c34aad7b9bcb7674f
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-03 02:11:17 +0000

    more dist.ini tuning, fixed version 

  Change: 8a6d6122292be9ee4e97345fc5db94a3c88d7238
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-03 02:00:35 +0000

    v0.1 

----------------------------------------
version 0.1 at 2013-05-02 13:24:33 +0000
----------------------------------------

  Change: 0226911bc2f0b3fde01a40bba82bc5e70a61f82b
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-03 01:24:33 +0000

    fixed typo in dist.ini, added README.pod 

  Change: 817341dd46b5151dafa9f3562da1b18d543bc5d5
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-02 17:40:54 +0000

    added github config 

  Change: e847c0f8e01400611d7e46c5b673195ca90482bf
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-02 16:30:30 +0000

    wrote proper tests 

  Change: 99650c72b6286ba3d266ded7841f9169f9ba8bec
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-02 16:30:08 +0000

    do not reset tax after calculate(), return only $item that was added,
    add default quantity of 1 unless given 

  Change: 2dca60e89d18c9ffcabb4051d13c92d0fc6bddea
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-02 15:12:16 +0000

    added non-working test 

  Change: 1f75a0e23751eb6dcdae4a8b636b87867c0f1e36
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-02 15:11:11 +0000

    perl critic fix 

  Change: dcc77a03f69398d2fef16c6c59ca23e7e02154ba
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2013-05-02 14:15:36 +0000

    initial release 

================
End of releases.
================


Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Basket/Cart calculation library with support for currency conversion, discounts and tax",
  "AUTHOR" => "Tobias Kirschstein <lev\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Basket-Calc",
  "EXE_FILES" => [],
  "LICENSE" => "bsd",
  "NAME" => "Basket::Calc",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Finance::Currency::Convert::Yahoo" => 0,
    "Mouse" => 0,
    "Scalar::Util" => 0,
    "experimental" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => 0
  },
  "VERSION" => "0.4",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Basket/Calc.pm
package Basket::Calc;

use 5.010;
use Mouse;
use experimental 'smartmatch';

# ABSTRACT: Basket/Cart calculation library with support for currency conversion, discounts and tax

our $VERSION = '0.4'; # VERSION

use Scalar::Util qw(looks_like_number);
use Finance::Currency::Convert::Yahoo;
use Carp;


has 'debug' => (
    is      => 'rw',
    isa     => 'Bool',
    trigger => \&_set_debug,
    lazy    => 1,
    default => sub { 0 },
);

has 'items' => (
    is      => 'rw',
    isa     => 'ArrayRef',
    clearer => 'empty_items',
);

has 'discount' => (
    is      => 'rw',
    isa     => 'HashRef',
    clearer => 'no_discount',
);


has 'currency' => (
    is       => 'rw',
    isa      => 'Str',
    required => 1,
);


has 'tax' => (
    is      => 'rw',
    isa     => 'Num',
    lazy    => 1,
    default => sub { 0 },
);


sub add_item {
    my ($self, $item) = @_;

    # make sure the input is sane
    unless (ref $item eq 'HASH') {
        carp "parameter has to be a HASHREF";
        return;
    }

    foreach my $key ('price') {
        unless (exists $item->{$key} and $item->{$key}) {
            carp "$key missing";
            return;
        }
    }

    unless (looks_like_number($item->{price})) {
        carp "'price' is not a number";
        return;
    }

    # calculate amount from quantity and price
    if (exists $item->{quantity}) {
        if (!looks_like_number($item->{quantity}) || ($item->{quantity} < 0)) {
            carp "'quantity' is not a number or smaller than 0";
            return;
        }

        $item->{amount} = $item->{price} * $item->{quantity};
    }
    else {
        $item->{amount}   = $item->{price};
        $item->{quantity} = 1;
    }

    $item->{currency} = $self->currency
        unless (exists $item->{currency} and $item->{currency});

    # convert currency if needed
    if ($item->{currency} ne $self->currency) {
        my $amount =
            Finance::Currency::Convert::Yahoo::convert($item->{amount},
            $item->{currency}, $self->currency);

        unless ($amount) {
            carp "could not get "
                . $item->{amount} . " "
                . $item->{currency}
                . " converted to "
                . $self->currency;
            return;
        }

        $item->{orig_amount}   = $item->{amount};
        $item->{orig_currency} = $item->{currency};

        $item->{amount}   = $amount;
        $item->{currency} = $self->currency;
    }

    print __PACKAGE__ . ' added item: ' . join(' ', %$item) . $/
        if $self->debug;

    $self->items([ @{ $self->items || [] }, $item ]);

    return $item;
}


sub add_discount {
    my ($self, $discount) = @_;

    # make sure the input is sane
    unless (ref $discount eq 'HASH') {
        carp "parameter has to be a HASHREF";
        return;
    }

    foreach my $key ('type', 'value') {
        unless (exists $discount->{$key} and $discount->{$key}) {
            carp "'$key' missing";
            return;
        }
    }

    unless ($discount->{type} =~ m/^(percent|amount)$/x) {
        carp "'type' has to be either percent, or amount";
        return;
    }

    unless (looks_like_number($discount->{value})) {
        carp "'value' is not a number";
        return;
    }

    given ($discount->{type}) {
        when ('percent') {
            if ($discount->{value} <= 0 or $discount->{value} > 1) {
                carp "'percent' has to be a decimal between 0 and 1";
                return;
            }
        }
        when ('amount') {
            $discount->{currency} = $self->currency
                unless exists $discount->{currency};

            # convert currency if needed
            if ($discount->{currency} ne $self->currency) {
                my $amount = Finance::Currency::Convert::Yahoo::convert(
                    $discount->{value}, $discount->{currency}, $self->currency);

                unless ($amount) {
                    carp "could not get "
                        . $discount->{value} . " "
                        . $discount->{currency}
                        . " converted to "
                        . $self->currency;
                    return;
                }

                $discount->{orig_value}    = $discount->{value};
                $discount->{orig_currency} = $discount->{currency};

                $discount->{value}    = $amount;
                $discount->{currency} = $self->currency;
            }
        }
    }

    print __PACKAGE__ . ' added discount: ' . join(' ', %$discount) . $/
        if $self->debug;

    $self->discount($discount);

    return $self->discount;
}


sub calculate {
    my ($self) = @_;

    unless ($self->items) {
        carp "no items added";
        return;
    }

    my $total = {
        value      => 0,
        net        => 0,
        tax_amount => 0,
        discount   => 0,
    };

    print __PACKAGE__ . " -- calculating totals --\n" if $self->debug;

    # calculate net
    foreach my $item (@{ $self->items }) {
        print __PACKAGE__ . ' item: ' . join(' ', %$item) . $/ if $self->debug;

        $total->{net} += $item->{amount};
    }

    my $original_net = $total->{net};

    # apply discounts
    if ($self->discount) {
        print __PACKAGE__ . ' discount: ' . join(' ', %{ $self->discount }) . $/
            if $self->debug;

        given ($self->discount->{type}) {
            when ('percent') {
                $total->{net} *= (1 - $self->discount->{value});
            }
            when ('amount') {
                $total->{net} = $total->{net} - $self->discount->{value};
                $total->{net} = 0 if $total->{net} < 0;
            }
        }
    }

    # calculate tax
    $total->{tax_amount} = $total->{net} * $self->tax;
    $total->{value}      = $total->{net} + $total->{tax_amount};
    $total->{discount}   = $original_net - $total->{net};

    # proper rounding and formatting
    $total->{$_} = _round($total->{$_}) for keys %$total;

    # remind what the currency is that was requested
    $total->{currency} = $self->currency;

    print __PACKAGE__ . ' total: ' . join(' ', %$total) . $/ if $self->debug;

    $self->empty_items;
    $self->no_discount;

    return $total;
}

sub _set_debug {
    my ($self, $value, $some) = @_;

    $Finance::Currency::Convert::Yahoo::CHAT = $value;

    return;
}

sub _round {
    my ($float) = @_;

    # some stupid perl versions on some platforms can't round correctly and i
    # don't want to use more modules
    $float += 0.001 if ($float =~ m/\.[0-9]{2}5/);

    return sprintf('%.2f', sprintf('%.10f', $float)) + 0;
}


1;    # End of Basket::Calc

__END__

=pod

=head1 NAME

Basket::Calc - Basket/Cart calculation library with support for currency conversion, discounts and tax

=head1 VERSION

version 0.4

=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use Basket::Calc;
    use Data::Dump 'dump';

    my $basket = Basket::Calc->new(debug => 1, currency => 'NZD', tax => .15);
    $basket->add_item({ price => 14.90, currency => 'USD', quantity => 2 });
    $basket->add_item({ price => 59, currency => 'EUR'});
    $basket->add_item({ price => 119.15, currency => 'JPY' });
    
    $basket->add_discount({ type => 'percent', value => .2 });
    # or
    $basket->add_discount({ type => 'amount', value => 15, currency => 'HKD' });
    
    print dump $basket->calculate;

=head1 ATTRIBUTES

=head2 debug

=head2 currency

=head2 tax

=head1 SUBROUTINES/METHODS

=head2 add_item

=head2 add_discount

=head2 calculate

=head1 BUGS

Please report any bugs or feature requests on GitHub's issue tracker L<https://github.com/nupfel/Basket-Calc/issues>.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Basket::Calc

You can also look for information at:

=over 4

=item * GitHub repository

L<https://github.com/nupfel/Basket-Calc>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Basket-Calc>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Basket-Calc>

=back

=head1 ACKNOWLEDGEMENTS

=over 4

=item * Lenz Gschwendtner (@norbu09), for being an awesome mentor and friend.

=back

=head1 AUTHOR

Tobias Kirschstein <lev@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by Tobias Kirschstein.

This is free software, licensed under:

  The (three-clause) BSD License

=cut
